# Data Cleaning: DPWH Flood Control Projects Dataset
Data Source: Kaggle ([DPWH Flood Control Projects](https://www.kaggle.com/datasets/bwandowando/dpwh-flood-control-projects))

---
## Import libraries:

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Load dataset:

In [2]:
df = pd.read_csv("/Users/saadult/dpwh-flood-control-projects/data/dpwh_flood_control_projects_raw.csv")

---
## Initial dataset inspection:

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9855 entries, 0 to 9854
Data columns (total 22 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   MainIsland                  9855 non-null   object 
 1   Region                      9855 non-null   object 
 2   Province                    9855 non-null   object 
 3   LegislativeDistrict         9855 non-null   object 
 4   Municipality                9189 non-null   object 
 5   DistrictEngineeringOffice   9855 non-null   object 
 6   ProjectId                   9855 non-null   object 
 7   ProjectName                 9855 non-null   object 
 8   TypeOfWork                  9855 non-null   object 
 9   FundingYear                 9855 non-null   int64  
 10  ContractId                  9855 non-null   object 
 11  ApprovedBudgetForContract   9855 non-null   object 
 12  ContractCost                9855 non-null   object 
 13  ActualCompletionDate        9855 

The dataset contain 22 columns/features and 9,855 rows/values. The values in the datasets are combinations of texts (objects) and numbers (integer and float). For columns such as `ApprovedBudgetForContract`, and `ContractCost`, values are in object format they need to be converted to numeric for proper analysis. Same with columns `ActualCompletionDate` and `StartDate`, values are currently in object format, need to convert to date format.

In [4]:
# Display all columns in the dataframe
pd.set_option('display.max_columns', None)

# Prevent column content from being truncated
pd.set_option('display.max_colwidth', None)

df.head()

,MainIsland,Region,Province,LegislativeDistrict,Municipality,DistrictEngineeringOffice,ProjectId,ProjectName,TypeOfWork,FundingYear,ContractId,ApprovedBudgetForContract,ContractCost,ActualCompletionDate,Contractor,ContractorCount,StartDate,ProjectLatitude,ProjectLongitude,ProvincialCapital,ProvincialCapitalLatitude,ProvincialCapitalLongitude
0,Luzon,Cordillera Administrative Region,Apayao,APAYAO (LEGISLATIVE DISTRICT),NaN,Apayao 2nd District Engineering Office,P00526268LZ,"Construction of Poblacion Flood Control Structure along Apayao River, Sta. 136 + 247.68 - Sta. 136 + 752.68 (L/S), Calanasan, Apayao",Construction of Flood Mitigation Structure,2021,21PC0018,96500000,92692943.8,2022-10-03,ASC CONSTRUCTION & CONCRETE PRODUCTS,1,2021-03-18,18.25091,121.047570,Kabugao,18.0215,121.1716
1,Luzon,Cordillera Administrative Region,Apayao,APAYAO (LEGISLATIVE DISTRICT),NaN,Apayao 2nd District Engineering Office,P00526368LZ,"Construction of Cabicungan Flood Control Structure, Sta. 13 + 580 - Sta. 13 + 605 (L/S), Sta 13+180-Sta 13+280 (L/S), Sta 13+460-Sta 13+510 (R/S), Sta 13+230-Sta 13+430 (R/S), Calanasan, Apayao",Construction of Flood Mitigation Structure,2021,21PC0019,49000000,47521251.3,2022-11-24,ASC CONSTRUCTION & CONCRETE PRODUCTS,1,2021-12-01,18.51180,121.078290,Kabugao,18.0215,121.1716
2,Luzon,Cordillera Administrative Region,Apayao,APAYAO (LEGISLATIVE DISTRICT),NaN,Apayao 2nd District Engineering Office,P00526369LZ,"Construction of Sipa-San Mariano Flood Control Structure along Apayao River, Sta. 25 + 380 - Sta. 25 + 700 (L/S), Sta. Marcela, Apayao",Construction of Flood Mitigation Structure,2021,21PC0010,47423180,45942583.67,2022-11-28,ASC CONSTRUCTION & CONCRETE PRODUCTS,1,2021-12-01,18.27023,121.420220,Kabugao,18.0215,121.1716
3,Luzon,Cordillera Administrative Region,Apayao,APAYAO (LEGISLATIVE DISTRICT),NaN,Apayao 2nd District Engineering Office,P00561270LZ,"Construction of Sta. Filomena Floodway/ Flood Control Structure, Sta. Filomena, Calanasan, Apayao",Construction of Flood Mitigation Structure,2021,21PC0067,9900000,9900000,2022-12-15,ASC CONSTRUCTION & CONCRETE PRODUCTS,1,2021-12-01,18.53566,121.080200,Kabugao,18.0215,121.1716
4,Luzon,Cordillera Administrative Region,Apayao,APAYAO (LEGISLATIVE DISTRICT),NaN,Apayao 2nd District Engineering Office,P00561652LZ,"Construction of Sinigpit Flood Control, Calanasan, Apayao",Construction of Flood Mitigation Structure,2021,21PC0034,4950000,4950000,2022-09-23,ALEPHNORTH PEAK CONSTRUCTION AND DEVELOPMENT CO.,1,2021-12-01,18.52723,121.096812,Kabugao,18.0215,121.1716


In [5]:
df.describe().round(2)

,FundingYear,ContractorCount,ProjectLatitude,ProjectLongitude,ProvincialCapitalLatitude,ProvincialCapitalLongitude
count,9855.00,9855.00,9855.00,9855.00,9855.00,9855.00
mean,2022.77,1.14,13.20,122.25,13.22,122.26
std,0.92,0.34,3.15,1.77,3.12,1.78
min,2018.00,1.00,5.77,117.06,6.05,118.74
25%,2022.00,1.00,10.72,120.91,10.68,120.81
50%,2023.00,1.00,14.27,121.44,14.28,121.44
75%,2023.00,1.00,15.16,123.72,15.03,123.85
max,2025.00,3.00,20.46,126.57,20.45,126.22


---
## Data validation:

In [6]:
df.isnull().sum()

MainIsland                      0
Region                          0
Province                        0
LegislativeDistrict             0
Municipality                  666
DistrictEngineeringOffice       0
ProjectId                       0
ProjectName                     0
TypeOfWork                      0
FundingYear                     0
ContractId                      0
ApprovedBudgetForContract       0
ContractCost                    0
ActualCompletionDate            0
Contractor                      0
ContractorCount                 0
StartDate                       0
ProjectLatitude                 0
ProjectLongitude                0
ProvincialCapital               0
ProvincialCapitalLatitude       0
ProvincialCapitalLongitude      0
dtype: int64

There are missing values under the `Municipality` column. The values in this column is very specific because it points to a specific place. It is not advisable to fill the missing values based on what is available so I will leave it as is for now.

In [7]:
df.duplicated().sum()

np.int64(0)

The are no duplicate values.

---
## Data cleaning:
### Converting `ApprovedBudgetForContract` and `ContractCost` to float:

In [8]:
# Create a copy of the original dataframe
df_clean = df.copy()

In [9]:
# Convert ApprovedBudgetForContract to float
df_clean['ApprovedBudgetForContract'] = pd.to_numeric(
    df_clean['ApprovedBudgetForContract'],
    errors='coerce'
)

# Convert ContractCost to float
df_clean['ContractCost'] = pd.to_numeric(
    df_clean['ContractCost'],
    errors='coerce'
)

In [10]:
# Confirm successful conversion
df_clean[['ApprovedBudgetForContract', 'ContractCost']].dtypes

ApprovedBudgetForContract    float64
ContractCost                 float64
dtype: object

In [11]:
# Count missing values after conversion
myca_budget_count = df_clean['ApprovedBudgetForContract'].isna().sum()
myca_cost_count = df_clean['ContractCost'].isna().sum()

print(f"ApprovedBudgetForContract missing after conversion: {myca_budget_count}")
print(f"ContractCost missing after conversion: {myca_cost_count}")

ApprovedBudgetForContract missing after conversion: 157
ContractCost missing after conversion: 28


In [12]:
# Percentage of affected records
(
    df_clean[['ApprovedBudgetForContract', 'ContractCost']]
    .isna()
    .mean()
    .mul(100)
    .round(2)
)

ApprovedBudgetForContract    1.59
ContractCost                 0.28
dtype: float64

During data cleaning, the `ApprovedBudgetForContract` and `ContractCost` fields were converted from object to numeric data types. A review of conversion errors revealed that some records contained MYCA (Multi-Year Contract Authority) references instead of monetary values. These entries were converted to NaN using pd.to_numeric(errors='coerce'). Approximately 157 budget records and 28 contract cost records were affected and retained as missing values for subsequent analysis. All 28 invalid contract cost records also had invalid approved budget values, indicating a consistent data-entry pattern rather than isolated errors.

### Converting `ActualCompletionDate` and `StartDate` to datetime format:

In [30]:
# Convert ActualCompletionDate and StartDate to datetime format
date_columns = ["ActualCompletionDate", "StartDate"]

for col in date_columns:
    df_clean[col] = pd.to_datetime(df_clean[col], errors="coerce")

# Verify the conversion
df_clean[date_columns].dtypes

ActualCompletionDate    datetime64[ns]
StartDate               datetime64[ns]
dtype: object

Converted ActualCompletionDate and StartDate to datetime format to enable date-based analysis, filtering, and project timeline calculations.

---
## One final check:

In [33]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9855 entries, 0 to 9854
Data columns (total 22 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   MainIsland                  9855 non-null   object        
 1   Region                      9855 non-null   object        
 2   Province                    9855 non-null   object        
 3   LegislativeDistrict         9855 non-null   object        
 4   Municipality                9189 non-null   object        
 5   DistrictEngineeringOffice   9855 non-null   object        
 6   ProjectId                   9855 non-null   object        
 7   ProjectName                 9855 non-null   object        
 8   TypeOfWork                  9855 non-null   object        
 9   FundingYear                 9855 non-null   int64         
 10  ContractId                  9855 non-null   object        
 11  ApprovedBudgetForContract   9698 non-null   float64     

In [32]:
df_clean.describe().round(2)

,FundingYear,ApprovedBudgetForContract,ContractCost,ActualCompletionDate,ContractorCount,StartDate,ProjectLatitude,ProjectLongitude,ProvincialCapitalLatitude,ProvincialCapitalLongitude
count,9855.00,9.698000e+03,9.827000e+03,9855,9855.00,9855,9855.00,9855.00,9855.00,9855.00
mean,2022.77,5.671067e+07,5.552457e+07,2023-09-26 07:49:19.999999744,1.14,2023-01-21 05:29:21.095890432,13.20,122.25,13.22,122.26
min,2018.00,1.158000e+06,1.154359e+06,2022-07-01 00:00:00,1.00,2018-02-26 00:00:00,5.77,117.06,6.05,118.74
25%,2022.00,2.744000e+07,2.450000e+07,2022-12-24 00:00:00,1.00,2022-03-25 00:00:00,10.72,120.91,10.68,120.81
50%,2023.00,4.900000e+07,4.899158e+07,2023-09-23 00:00:00,1.00,2023-02-24 00:00:00,14.27,121.44,14.28,121.44
75%,2023.00,8.685000e+07,8.523367e+07,2024-05-24 00:00:00,1.00,2023-08-30 00:00:00,15.16,123.72,15.03,123.85
max,2025.00,3.105763e+08,2.895000e+08,2025-05-30 00:00:00,3.00,2025-03-14 00:00:00,20.46,126.57,20.45,126.22
std,0.92,4.009798e+07,3.949880e+07,NaN,0.34,NaN,3.15,1.77,3.12,1.78


---
## Save cleaned dataset:

In [35]:
# Save cleaned dataset to the data folder
df_clean.to_csv('data/dpwh_flood_control_projects_clean.csv', index=False)

---
## Summary:
**Data Structure**
* The dataset contains a mix of categorical, numerical, and date-related fields.
* Several important fields, including `ApprovedBudgetForContract` and `ContractCost`, were initially stored as object data types and required conversion to numeric format for analysis.
* Converted StartDate and ActualCompletionDate to date format to support time-based analysis and project timeline evaluation.

**Missing Values**
* Missing values were identified in the Municipality column.
* Because municipality names represent specific geographic locations, no imputation was performed to avoid introducing inaccurate information.

**Duplicate Records**
* No duplicate records were found in the dataset, indicating a high level of record uniqueness and integrity.

**Data Quality Issues**
* During numeric conversion, some records in `ApprovedBudgetForContract` and `ContractCost` contained text-based MYCA (Multi-Year Contract Authority) references rather than monetary values.
* These entries were converted to NaN using pd.to_numeric(errors='coerce').
* A total of:
    * 157 records were affected in `ApprovedBudgetForContract`
    * 28 records were affected in `ContractCost`
* All 28 invalid contract cost records were also associated with invalid approved budget values, suggesting a consistent reporting pattern rather than random data-entry errors.